### PR TITLE
DR-1147 part 3: Allow test scripts to call API clients other than just Data Repo.

### PR DESCRIPTION
**Pull request by marikomedlock**
Main change is to make it easier to use other client objects, not just Data Repo, with the test user's credentials.
- Changed the Test Script user journey method to accept a TestUser specification instead of a Data Repo API client.
- Added methods to utils classes to fetch client objects for a given test user. Client objects available so far are: Data Repo, SAM, BigQuery.
- Added a new test script DatasetCustodianPermissions that uses all three available client objects to check permissions on a dataset for various test users.

Other changes:
- Moved often-copied code to create/delete a profile and dataset into a common base class SimpleDataset, and changed the various TestScript classes to extend it.
- Added code to SimpleDataset base class to check all the test users to find a steward, instead of just using the first user in the list.
- Changed the isFailure flag to also be true if numExceptions != 0.
- Reorganized test config JSON files into sub-directories.
- Added an environment variable TEST_RUNNER_SERVER_SPECIFICATION_FILE to optionally override the server specification for a given run.
- Fixed a bug that Shelby reported where the different thread pools for each test script were not being run in parallel. Changed the thread submission to the pool to be asynchronous.

### CODE REVIEW

File: datarepo-clienttests/src/main/java/runner/TestScriptResult.java [hunk 1]
```diff
@@ -67,7 +67,8 @@ private void calculateStatistics() {
     summary.meanElapsedTimeMS /= userJourneyResults.size();
     summary.totalRun = userJourneyResults.size();
 
-    summary.isFailure = summary.numCompleted < summary.totalRun;
+    summary.isFailure =
+        (summary.numCompleted < summary.totalRun) || (summary.numExceptionsThrown > 0);
```
Comment (RoriCremer) [inline datarepo-clienttests/src/main/java/runner/TestScriptResult.java:72]: awesome!

File: datarepo-clienttests/README.md [hunk 2]
```diff
@@ -76,10 +76,18 @@ Find a test configuration to execute. Each configuration is a JSON file in the r
...
 
 `./gradlew run --args="suites/BasicSmoke.json"`
 
+The environment variable TEST_RUNNER_SERVER_SPECIFICATION_FILE optionally overrides the server configuration specified
+by either the test suite or configuration. The server specification is determined by the following, in order:
+  1. environment variable
+  2. test suite server property (if running a test suite)
+  3. test configuration server property
+  
+`TEST_RUNNER_SERVER_SPECIFICATION_FILE="mmdev.json" ./gradlew run --args="configs/basicexamples/LookupDataset.json"`
```
Comment (snf2ye) [inline datarepo-clienttests/README.md:89]: This is great - thank you!

File: datarepo-clienttests/src/main/java/runner/TestRunner.java [hunk 3]
```diff
@@ -178,18 +160,19 @@ void executeTestConfigurationNoGuaranteedCleanup() throws Exception {
...
-            new UserJourneyThread(testScript, testScriptSpecification.description, apiClient));
+        Future<UserJourneyResult> userJourneyFuture =
+            threadPool.submit(
+                new UserJourneyThread(testScript, testScriptSpecification.description, testUser));
+        userJourneyFutures.add(userJourneyFuture);
       }
 
       // TODO: support different patterns of kicking off user journeys. here they're all queued at
       // once
-      List<Future<UserJourneyResult>> userJourneyFutures = threadPool.invokeAll(userJourneyThreads);
```
Comment (snf2ye) [inline datarepo-clienttests/src/main/java/runner/TestRunner.java:192]: Ok, great! This should cover the issue that I was seeing.

File: datarepo-clienttests/src/main/java/runner/config/TestConfiguration.java [hunk 4]
```diff
@@ -80,10 +94,6 @@ public void validate() {
     for (TestScriptSpecification testScript : testScripts) {
       testScript.validate();
       if (isFunctional) {
-        if (testScript.totalNumberToRun > 1) {
-          throw new IllegalArgumentException(
-              "For a functional test script, the total number to run should be 1.");
-        }
```
Comment (snf2ye) [inline datarepo-clienttests/src/main/java/runner/config/TestConfiguration.java:86]: Why is this removed?

File: datarepo-clienttests/src/main/java/testscripts/baseclasses/SimpleDataset.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+public class SimpleDataset extends runner.TestScript {
+  private static final Logger logger = LoggerFactory.getLogger(SimpleDataset.class);
+
+  protected TestUserSpecification datasetCreator;
+  protected BillingProfileModel billingProfileModel;
+  protected DatasetSummaryModel datasetSummaryModel;
+
+  public void setup(List<TestUserSpecification> testUsers) throws Exception {
+    // pick the a user that is a Data Repo steward to be the dataset creator
+    datasetCreator = SAMUtils.findTestUserThatIsDataRepoSteward(testUsers, server);
```
Comment (snf2ye) [inline datarepo-clienttests/src/main/java/testscripts/baseclasses/SimpleDataset.java:26]: Ok, sanity checking that I understand these changes:
* The first user w/ steward permissions we find from our list of user to perform steward tasks, particularly in the setup and cleanup methods
* All of the users defined in the test config are cycled through to perform the "user tasks", like api calls, in the user journey

I vaguely remember there being some desire to spread the tasks/permissions between test users. 
Is it at all worth it to make these user selections a more random? According to my understanding above, I think the first  user in the list would be used a lot more. Just throwing this out there - probably not worth it.

File: datarepo-clienttests/src/main/java/testscripts/BulkLoad.java [hunk 6]
```diff
@@ -31,26 +34,25 @@ public void setParameters(List<String> parameters) throws Exception {
...
 
-    // wait for the job to complete and print out results
-    bulkLoadUtils.getAndDisplayResults(repositoryApi, bulkLoadArrayJobResponse);
-  }
+    // wait for the job to complete and print out results to debug log level
+    BulkLoadResultModel loadSummary =
+        BulkLoadUtils.getAndDisplayResults(repositoryApi, bulkLoadArrayJobResponse);
 
-  public void cleanup(Map<String, ApiClient> apiClients) throws Exception {
-    KubernetesClientUtils.changeReplicaSetSizeAndWait(1);
```
Comment (snf2ye) [inline datarepo-clienttests/src/main/java/testscripts/BulkLoad.java:53]: Do we need to reset the replica set size?

File: datarepo-clienttests/src/main/java/utils/KubernetesClientUtils.java [hunk 7]
```diff
@@ -258,7 +259,7 @@ public static void deleteRandomPod() throws ApiException, IOException {
                 pod ->
                     deploymentComponentLabel.equals(
                         pod.getMetadata().getLabels().get(componentLabel)))
-            .skip((int) (podCount * Math.random()))
+            .skip(new Random().nextInt((int) podCount))
```
Comment (snf2ye) [inline datarepo-clienttests/src/main/java/utils/KubernetesClientUtils.java:262]: Thanks for fixing this bug!

**Review by snf2ye — APPROVED** [review]
Looks great!

**Review by fboulnois — COMMENTED** [review]
I took a quick look and I didn't see anything unusual. I also think the refactoring is useful.

File: datarepo-clienttests/src/main/java/runner/config/TestConfiguration.java (hunk 4 above)
Comment (marikomedlock) [reply datarepo-clienttests/src/main/java/runner/config/TestConfiguration.java:86]: Because I realized that some functional tests may want to run as more than one test user.

File: datarepo-clienttests/src/main/java/testscripts/baseclasses/SimpleDataset.java (hunk 5 above)
Comment (marikomedlock) [reply datarepo-clienttests/src/main/java/testscripts/baseclasses/SimpleDataset.java:26]: Yes, your summary is right. Regarding the suggestion to make the dataset creator selection more random, I think that's a good idea. I'll update the code.

File: datarepo-clienttests/src/main/java/testscripts/BulkLoad.java (hunk 6 above)
Comment (marikomedlock) [reply datarepo-clienttests/src/main/java/testscripts/BulkLoad.java:53]: I think we don't need to because we already set the replica set size at the beginning of every test run. There's no real reason to "put things back to normal" because we wipe out the deployment before every test run anyway.

File: datarepo-clienttests/src/main/java/utils/KubernetesClientUtils.java (hunk 7 above)
Comment (marikomedlock) [reply datarepo-clienttests/src/main/java/utils/KubernetesClientUtils.java:262]: Well, not really a bug -- your version is easier to read in my opinion. But spotbugs was upset.

**Review by fboulnois — APPROVED** [review]
latest changes lgtm!
